### ISSUE TITLE
[AspireDeployment] Project Build's errors occurring on JSON RPC invocation should be better reported back to client

### ISSUE THREAD

**Issue by anvillan** [Azure/azure-dev#3316]
When a JSON RPC request is done to the `azd vs-server`, and the project needs to be built, any build error is reported as follow:

`StreamJsonRpc.RemoteInvocationException: failed to load app host manifest: failed to load app host manifest: generating app host manifest: dotnet run --publisher manifest on project 'C:\Users\anvillan\source\repos\AspireApp2\AspireApp2.AppHost\AspireApp2.AppHost.csproj' failed: exit code: 1, stdout: Building...
C:\Users\anvillan\source\repos\AspireApp2\AspireApp2.ServiceDefaults\AspireApp2.ServiceDefaults.csproj : error NU1102: Unable to find package Microsoft.Extensions.ServiceDiscovery with version (>= 8.0.0-preview.3.24105.21) 
...
, stderr: 
**The build failed. Fix the build errors and run again.**

   at StreamJsonRpc.JsonRpc.<InvokeCoreAsync>d__154`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at Microsoft.VisualStudio.ConnectedServices.Azure.AzDev.AzdAspireService.<>c__DisplayClass4_0.<<GetAspireHostAsync>b__0>d.MoveNext() in D:\WebTools\src\ConnectedServices\ConnectedServices.Package2\Azure\AzDev\Impl\AzdAspireService.cs:line 37
--- End of stack trace from previous location where exception was thrown ---
...`

Certainly, a user looking at the VS IDE `AZD Cli` output panel may spot the line stating a Build error occurred, but as the error is treated as any other failure (in the `STDERR`) it might also be overlooked.

Is there a way to _better_ communicate a build error as such so that VS IDE can _take it_ and show it in a more prominently way?

**Comment by rajeshkamal5050** [comment]
Tirage update: Not a must-fix item for `VS 17.10 Preview 2`. Moving to next milestone.

@anvillan @weikanglim @abpiskunov @ellismg

**Comment by weikanglim** [comment]
@anvillan 

Happy to provide structured errors and error codes once we have ideas of how we'd want to present this in VS.

As a short-term improvement while we're still writing to the output terminal, I wonder for cases of [RemoteInvocationException](https://github.com/microsoft/vs-streamjsonrpc/blob/main/doc/exceptions.md#client-side-concerns)*, on the VS client, could inspect the `error.message` contained within the exception and only print out `error.message` to the terminal.

This would end up working nicely. For example, the `error.message` would avoid the client callstack, and more easily highlight "not logged in".

![image (2)](https://github.com/Azure/azure-dev/assets/2322434/ae0626f8-ce99-4cad-bf0c-771160801c5d)

*Footnote: Technically, RemoteInvocationException could include other protocol / client errors, but just logging the message would be clear enough for debugging.

**Comment by anvillan** [comment]
@weikanglim 

As we discussed, let's ...

 - Use the `IObserver<ProgressMessage>` to send back a structured error message when any of the exposed endpoints generate such error; the proposal of taking the `error.message` sounds good for starters. Let's begin with the most visible invocation,`EnvironmentService.DeployAsync()`, though we should do it for all our API endpoints.
 
- In the `azd.exe` process execution nothing should change from our current implementation: we should keep sending all the `STDOUT`/`STDERR` as we currently do.

CC @abpiskunov

**Comment by weikanglim** [comment]
@anvillan 

We can send the RPC call failure error message in `EnvironmentService.DeployAsync()` using Observer notifications immediately to unblock. In VS, this ends up sending the error message to the "Build" output, while "Azure Developer CLI" output would have the same message from the RPC call failure. This works nicely.

For operations other than `EnvironmentService.DeployAsync()`, if we're not separating the "observer output" from the "rpc log" output, this would end up printing the same message twice. If we're thinking there's always two different output terminals, this problem goes away. If not, we may need to make some client changes here.

**Comment by anvillan** [comment]
@weikanglim 

Understood. Let's start only with `EnvironmentService.DeployAsync()` as you correctly point out, it is the only one where we have a _separate_ output panel. 

I'd say we keep in mind that if we have the need to do the same operation(s), the code _ready_ to do so.

**Comment by weikanglim** [comment]
@anvillan I started thinking about this more over the weekend as I tried implementing this on the server side, and I think there are some issues we want to avoid:

- Constant churn between server-side and client-side handling (i.e. doing it for just one isn't going to work)
- Protocol errors between server and client would still be unreported (note that `IObserver` only works when the client<->server RPC handshake succeeds, and the server is sending back data in the nominal flow)

For those reasons, I started thinking what it'd look like if we just reported errors on the client-side. I submitted what an alternative proposal would look like in our internal Webtools repo. I hope we can discuss this, because I do prefer this approach for the reasons listed above. We can still use `IObserver` in the future to send back info/warning/error message streams, but we also want to handle RPC call failures in a nominal way so that the user is always given the right error context info when things do go wrong.
